File: AssetSync/src/AssetSync.Infrastructure/Api/IruService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using AssetSync.Core.Interfaces;
using AssetSync.Core.Models;
using AssetSync.Core.Services;

namespace AssetSync.Infrastructure.Api;

public class IruService : IIruService
{
    private const int PageSize = 300;
    private readonly string _baseUrl;
    private readonly Func<string> _getToken;
    private readonly IHttpClientFactory _httpClientFactory;

    public IruService(string baseUrl, Func<string> getToken, IHttpClientFactory httpClientFactory)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _getToken = getToken;
        _httpClientFactory = httpClientFactory;
    }

    public async Task<IReadOnlyList<Device>> GetDevicesAsync(CancellationToken cancellationToken = default)
    {
        var client = _httpClientFactory.CreateClient();
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _getToken());
        var list = new List<Device>();
        var offset = 0;

        while (true)
        {
            var url = $"{_baseUrl}/api/v1/devices?limit={PageSize}&offset={offset}";
            var response = await client.GetAsync(url, cancellationToken).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();
            var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            // Kandji returns an array of devices directly, or an object with "results"
            JsonElement devices;
            if (root.ValueKind == JsonValueKind.Array)
                devices = root;
            else if (root.ValueKind == JsonValueKind.Object && root.TryGetProperty("results", out var results) && results.ValueKind == JsonValueKind.Array)
                devices = results;
            else
                break;

            var pageCount = 0;
            foreach (var r in devices.EnumerateArray())
            {
                if (r.ValueKind != JsonValueKind.Object) continue;
                pageCount++;
                var serial = r.TryGetProperty("serial_number", out var sn) ? sn.GetString() ?? "" : "";
                var deviceName = r.TryGetProperty("device_name", out var dn) ? dn.GetString() : null;
                var model = r.TryGetProperty("model", out var m) ? m.GetString() : null;
                var osVersion = r.TryGetProperty("os_version", out var ov) ? ov.GetString() : null;
                var deviceId = r.TryGetProperty("device_id", out var did) ? did.GetString() : null;
                deviceId ??= r.TryGetProperty("id", out var id) ? id.GetString() : null;
                string? userEmail = null;
                if (r.TryGetProperty("user", out var user) && user.ValueKind == JsonValueKind.Object && user.TryGetProperty("email", out var em))
                    userEmail = em.GetString();
                list.Add(new Device
                {
                    NormalizedSerial = SerialNumberNormalizer.Normalize(serial),
                    SerialNumber = serial,
                    DeviceName = deviceName,
                    Model = model,
                    AssignedUserUpn = userEmail,
                    OsVersion = osVersion,
                    PlatformSource = "Iru",
                    IruDeviceId = deviceId,
                    DeviceType = r.TryGetProperty("device_family", out var df) ? df.GetString() : null
                });
            }

            // If we got fewer than PageSize results, we've reached the end
            if (pageCount < PageSize)
                break;

            offset += pageCount;
        }

        return list;
    }

    public async Task<bool> WriteBackAssetTagAsync(string iruDeviceId, string assetTag, CancellationToken cancellationToken = default)
    {
        try
        {
            var client = _httpClientFactory.CreateClient();
            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _getToken());
            var payload = JsonSerializer.Serialize(new { asset_tag = assetTag });
            var content = new StringContent(payload, Encoding.UTF8, "application/json");
            var response = await client.PatchAsync($"{_baseUrl}/api/v1/devices/{iruDeviceId}", content, cancellationToken).ConfigureAwait(false);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}

File: AssetSync-macOS/AssetSyncInfrastructure/Sources/AssetSyncInfrastructure/Api/IruService.swift
import Foundation
import AssetSyncCore

/// Fetches devices from the Iru (formerly Kandji) API with pagination.
public final class IruService: IruServiceProtocol, @unchecked Sendable {
    private static let pageSize = 300
    private let baseURL: String
    private let getToken: @Sendable () -> String
    private let session: URLSession

    public init(baseURL: String, getToken: @escaping @Sendable () -> String, session: URLSession = .shared) {
        self.baseURL = baseURL.trimmingSuffix("/")
        self.getToken = getToken
        self.session = session
    }

    public func getDevices() async throws -> [Device] {
        var devices: [Device] = []
        var offset = 0

        while true {
            var request = URLRequest(url: URL(string: "\(baseURL)/api/v1/devices?limit=\(Self.pageSize)&offset=\(offset)")!)
            request.setValue("Bearer \(getToken())", forHTTPHeaderField: "Authorization")

            let (data, _) = try await session.data(for: request)
            let json = try JSONSerialization.jsonObject(with: data)

            let array: [[String: Any]]
            if let arr = json as? [[String: Any]] {
                array = arr
            } else if let obj = json as? [String: Any], let results = obj["results"] as? [[String: Any]] {
                array = results
            } else {
                break
            }

            for r in array {
                let serial = r["serial_number"] as? String ?? ""
                let userEmail: String? = (r["user"] as? [String: Any])?["email"] as? String
                let deviceId = (r["device_id"] as? String) ?? (r["id"] as? String)
                let assetTag = (r["asset_tag"] as? String)?.trimmingCharacters(in: .whitespaces)

                var device = Device()
                device.normalizedSerial = SerialNumberNormalizer.normalize(serial)
                device.serialNumber = serial
                device.deviceName = r["device_name"] as? String
                device.model = r["model"] as? String
                device.assignedUserUpn = userEmail
                device.osVersion = r["os_version"] as? String
                device.platformSource = "Iru"
                device.iruDeviceId = deviceId
                device.deviceType = r["device_family"] as? String
                device.mdmAssetTag = (assetTag?.isEmpty ?? true) ? nil : assetTag
                devices.append(device)
            }

            if array.count < Self.pageSize { break }
            offset += array.count
        }

        return devices
    }

    public func writeBackAssetTag(iruDeviceId: String, assetTag: String) async throws -> Bool {
        var request = URLRequest(url: URL(string: "\(baseURL)/api/v1/devices/\(iruDeviceId)")!)
        request.httpMethod = "PATCH"
        request.setValue("Bearer \(getToken())", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["asset_tag": assetTag])

        let (_, response) = try await session.data(for: request)
        return (response as? HTTPURLResponse)?.statusCode.isSuccessful ?? false
    }
}

private extension Int {
    var isSuccessful: Bool { (200..<300).contains(self) }
}

private extension String {
    func trimmingSuffix(_ suffix: String) -> String {
        hasSuffix(suffix) ? String(dropLast(suffix.count)) : self
    }
}
